File: captchax/sdk/go/captchax/internal/http.go
package captchax

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type HTTPClient struct {
	baseURL    string
	timeout    time.Duration
	retryTimes int
	headers    map[string]string
	client     *http.Client
}

func NewHTTPClient(baseURL string, timeout time.Duration, retryTimes int) *HTTPClient {
	baseURL = strings.TrimSuffix(baseURL, "/")
	return &HTTPClient{
		baseURL:    baseURL,
		timeout:    timeout,
		retryTimes: retryTimes,
		headers: map[string]string{
			"Content-Type": "application/json",
			"Accept":       "application/json",
		},
		client: &http.Client{
			Timeout: timeout,
		},
	}
}

func (c *HTTPClient) SetHeader(key, value string) {
	c.headers[key] = value
}

func (c *HTTPClient) SetHeaders(headers map[string]string) {
	for k, v := range headers {
		c.headers[k] = v
	}
}

func (c *HTTPClient) buildURL(endpoint string) string {
	if strings.HasPrefix(endpoint, "http") {
		return endpoint
	}
	endpoint = strings.TrimPrefix(endpoint, "/")
	return fmt.Sprintf("%s/%s", c.baseURL, endpoint)
}

func (c *HTTPClient) request(ctx context.Context, method, endpoint string, body interface{}, deduplicationID string) ([]byte, *Error) {
	var lastErr error
	for attempt := 0; attempt <= c.retryTimes; attempt++ {
		req, err := c.buildRequest(ctx, method, endpoint, body, deduplicationID)
		if err != nil {
			return nil, NewError(err.Error())
		}

		resp, err := c.client.Do(req)
		if err != nil {
			lastErr = err
			if attempt < c.retryTimes {
				time.Sleep(time.Duration(1<<uint(attempt)) * time.Second)
				continue
			}
			return nil, NewError(fmt.Sprintf("request failed after %d attempts: %v", c.retryTimes+1, err))
		}
		defer resp.Body.Close()

		respBody, err := io.ReadAll(resp.Body)
		if err != nil {
			return nil, NewError(fmt.Sprintf("failed to read response: %v", err))
		}

		if resp.StatusCode >= 500 && attempt < c.retryTimes {
			lastErr = fmt.Errorf("server error: %d", resp.StatusCode)
			time.Sleep(time.Duration(1<<uint(attempt)) * time.Second)
			continue
		}

		if resp.StatusCode >= 400 {
			var apiResp APIResponse
			if err := json.Unmarshal(respBody, &apiResp); err == nil && apiResp.Message != "" {
				return nil, NewErrorWithCode(apiResp.Message, apiResp.Code, resp.StatusCode)
			}
			return nil, NewErrorWithCode(fmt.Sprintf("HTTP error: %d", resp.StatusCode), resp.StatusCode, resp.StatusCode)
		}

		return respBody, nil
	}

	return nil, NewError(fmt.Sprintf("request failed after %d attempts: %v", c.retryTimes+1, lastErr))
}

func (c *HTTPClient) buildRequest(ctx context.Context, method, endpoint string, body interface{}, deduplicationID string) (*http.Request, error) {
	urlStr := c.buildURL(endpoint)
	var reqBody io.Reader

	if body != nil && method != http.MethodGet {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		reqBody = bytes.NewReader(jsonBody)
	}

	req, err := http.NewRequestWithContext(ctx, method, urlStr, reqBody)
	if err != nil {
		return nil, err
	}

	for k, v := range c.headers {
		req.Header.Set(k, v)
	}

	if deduplicationID != "" {
		req.Header.Set("X-Deduplication-ID", deduplicationID)
	}

	return req, nil
}

func (c *HTTPClient) Get(ctx context.Context, endpoint string) ([]byte, *Error) {
	return c.request(ctx, http.MethodGet, endpoint, nil, "")
}

func (c *HTTPClient) Post(ctx context.Context, endpoint string, body interface{}, deduplicationID string) ([]byte, *Error) {
	return c.request(ctx, http.MethodPost, endpoint, body, deduplicationID)
}

func (c *HTTPClient) Put(ctx context.Context, endpoint string, body interface{}) ([]byte, *Error) {
	return c.request(ctx, http.MethodPut, endpoint, body, "")
}

func (c *HTTPClient) Delete(ctx context.Context, endpoint string) ([]byte, *Error) {
	return c.request(ctx, http.MethodDelete, endpoint, nil, "")
}

func (c *HTTPClient) parseResponse(body []byte, v interface{}) *Error {
	var apiResp APIResponse
	if err := json.Unmarshal(body, &apiResp); err != nil {
		return NewError(fmt.Sprintf("failed to parse response: %v", err))
	}

	if apiResp.Code != 0 && apiResp.Code != 200 {
		return NewErrorWithCode(apiResp.Message, apiResp.Code, 400)
	}

	if apiResp.Data != nil {
		dataBytes, err := json.Marshal(apiResp.Data)
		if err != nil {
			return NewError(fmt.Sprintf("failed to marshal data: %v", err))
		}
		if err := json.Unmarshal(dataBytes, v); err != nil {
			return NewError(fmt.Sprintf("failed to unmarshal data: %v", err))
		}
	}

	return nil
}

File: captchax/sdk/javascript/src/http.ts
import { CaptchaXError, CaptchaResponse } from './types';

const DEFAULT_TIMEOUT = 10000;
const DEFAULT_RETRY_TIMES = 3;

interface RequestOptions extends RequestInit {
  timeout?: number;
  retryTimes?: number;
  deduplicationId?: string;
}

export class HttpClient {
  private baseUrl: string;
  private defaultHeaders: Record<string, string>;
  private defaultTimeout: number;
  private defaultRetryTimes: number;

  constructor(baseUrl: string, timeout: number = DEFAULT_TIMEOUT, retryTimes: number = DEFAULT_RETRY_TIMES) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.defaultHeaders = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };
    this.defaultTimeout = timeout;
    this.defaultRetryTimes = retryTimes;
  }

  public setHeader(key: string, value: string): void {
    this.defaultHeaders[key] = value;
  }

  public setHeaders(headers: Record<string, string>): void {
    this.defaultHeaders = { ...this.defaultHeaders, ...headers };
  }

  private async fetchWithTimeout(
    url: string,
    options: RequestInit,
    timeout: number
  ): Promise<Response> {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const response = await fetch(url, {
        ...options,
        signal: controller.signal,
      });
      return response;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  private async request<T>(
    method: string,
    endpoint: string,
    body?: unknown,
    options: Partial<RequestOptions> = {}
  ): Promise<T> {
    const url = `${this.baseUrl}${endpoint}`;
    const timeout = options.timeout ?? this.defaultTimeout;
    const retryTimes = options.retryTimes ?? this.defaultRetryTimes;
    const deduplicationId = options.deduplicationId;

    const headers: Record<string, string> = { ...this.defaultHeaders };
    if (options.headers) {
      Object.assign(headers, options.headers);
    }

    const requestInit: RequestInit = {
      method,
      headers,
      ...options,
    };

    if (body && method !== 'GET') {
      requestInit.body = JSON.stringify(body);
    }

    let lastError: Error | null = null;

    for (let attempt = 0; attempt <= retryTimes; attempt++) {
      try {
        const requestHeaders = { ...headers };
        if (deduplicationId) {
          requestHeaders['X-Deduplication-ID'] = deduplicationId;
        }

        const response = await this.fetchWithTimeout(
          url,
          { ...requestInit, headers: requestHeaders },
          timeout
        );

        const data = await response.json() as CaptchaResponse<T>;

        if (!response.ok) {
          const errorMessage = data.message || `HTTP error: ${response.status}`;
          throw new CaptchaXError(
            errorMessage,
            data.code || response.status,
            response.status,
            data
          );
        }

        return data.data;
      } catch (error) {
        if (error instanceof CaptchaXError) {
          if (error.statusCode >= 500 && attempt < retryTimes) {
            lastError = error;
            await this.delay(Math.pow(2, attempt) * 100);
            continue;
          }
          throw error;
        }

        if (error instanceof Error) {
          if (error.name === 'AbortError') {
            throw new CaptchaXError('Request timeout', 408, 408);
          }

          if (attempt < retryTimes) {
            lastError = error;
            await this.delay(Math.pow(2, attempt) * 100);
            continue;
          }

          throw new CaptchaXError(
            `Network error: ${error.message}`,
            0,
            0,
            error
          );
        }

        lastError = new Error('Unknown error');
      }
    }

    throw new CaptchaXError(
      `Request failed after ${retryTimes + 1} attempts: ${lastError?.message}`,
      0,
      0,
      lastError
    );
  }

  private delay(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }

  public async get<T>(endpoint: string, options?: Partial<RequestOptions>): Promise<T> {
    return this.request<T>('GET', endpoint, undefined, options);
  }

  public async post<T>(
    endpoint: string,
    body: unknown,
    options?: Partial<RequestOptions>
  ): Promise<T> {
    return this.request<T>('POST', endpoint, body, options);
  }

  public async put<T>(
    endpoint: string,
    body: unknown,
    options?: Partial<RequestOptions>
  ): Promise<T> {
    return this.request<T>('PUT', endpoint, body, options);
  }

  public async delete<T>(
    endpoint: string,
    options?: Partial<RequestOptions>
  ): Promise<T> {
    return this.request<T>('DELETE', endpoint, undefined, options);
  }
}
